Adds initialization progress to profile API

kaituo opened the pull request:
*Issue #, if available:*
https://github.com/opendistro-for-elasticsearch/anomaly-detection/issues/148

*Description of changes:*
This PR adds init_progress to profile API. init_progress helps users track initialization percentage, needed shingles, and estimated time to go if the future data stream is continuous (no missing data).

Initialization percentage measures how far away we can observe RCF models emitting scores.  The implementation fetches the RCF model's total updates while the AD job is running and materializes the value to the newly added index .opendistro-anomaly-info. Total updates record the number of times this RCF model has been updated. 

Initialization percent is computed as x/128: 
* if total updates > 128, x = 128. Otherwise, x is the total updates 
* 128 is our output after the number in RCF. After observing 128 samples, RCF starts emitting scores.

Needed shingles are computed as 128 -x.  Estimated minutes to go is computed as needed shingles * detector interval

This PR also materializes the error message in the most recent run to speed up profile API's error fetching.

During each AD execution, we also check if a checkpoint is there (the result is saved and maintained as other AD states), if yes, we cold start immediately.

Testing done:
1. adds unit tests
2. run e2e tests to verify init_progress number makes sense.

Sample output:

```
kaituo@dev-dsk-kaituo-2a-74e3f324.us-west-2: ~/code/github/anomaly-detection
% curl -X GET "localhost:9200/_opendistro/_anomaly_detection/detectors/eT6_yXIBegfEFVlvLSF3/_profile?_all=true&pretty"
{
  "state" : "INIT",
  "models" : [
    {
      "model_id" : "eT6_yXIBegfEFVlvLSF3_model_rcf_2",
      "model_size_in_bytes" : 2228224,
      "node_id" : "4l_Pb26iRLWrFydpEhVBsA"
    },
    {
      "model_id" : "eT6_yXIBegfEFVlvLSF3_model_rcf_1",
      "model_size_in_bytes" : 2228224,
      "node_id" : "09xMkHzuSqOEIyn__syHUQ"
    },
    {
      "model_id" : "eT6_yXIBegfEFVlvLSF3_model_rcf_0",
      "model_size_in_bytes" : 2228224,
      "node_id" : "09xMkHzuSqOEIyn__syHUQ"
    },
    {
      "model_id" : "eT6_yXIBegfEFVlvLSF3_model_threshold",
      "node_id" : "09xMkHzuSqOEIyn__syHUQ"
    }
  ],
  "shingle_size" : 7,
  "coordinating_node" : "4l_Pb26iRLWrFydpEhVBsA",
  "total_size_in_bytes" : 6684672,
  "init_progress" : {
    "percentage" : "70%",
    "estimated_minutes_left" : 77,
    "needed_shingles" : 77
  }
}
[2020-06-18 16:31:32 PDT] [2020-06-18 23:31:32 UTC]
kaituo@dev-dsk-kaituo-2a-74e3f324.us-west-2: ~/code/github/anomaly-detection
% curl -X GET "localhost:9200/_opendistro/_anomaly_detection/detectors/eT6_yXIBegfEFVlvLSF3/_profile/init_progress,state,error"
{"state":"INIT","init_progress":{"percentage":"70%","estimated_minutes_left":77,"needed_shingles":77}}%
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

kaituo commented:
> Two general comments.
> 
>     1. Does additional detector info require an additional data index? An additional internal field in detector is a better option for it requires less resources (no new index or redundant data fields) and easier to manage (less consistency issues from multiple indices).
The idea is to put all state related to an detector to the info index.  As you said, additional detector info field will be added to the info index if required. 
 
>     2. After initialization is completed, the status is changed and collecting data for computing initializatio progress should stop. Collecting data and computing can be done on request instead.
I prefer to keep collecting after the status is changed:
First, we need extra logic to keep track of detector state.  For example, a detector is stopped and its init_progress should be recalculated (we don't know its models are deleted or not, its detector definition is updated or not).  Tracking the state is not trivial, and we should not put the burden on ModelManager or AnomalyResultTransportAction.
Second, tracking can help us debugging when we suspect whether the model is not working or not.

wnbts commented:
> > Two general comments.
> > ```
> > 1. Does additional detector info require an additional data index? An additional internal field in detector is a better option for it requires less resources (no new index or redundant data fields) and easier to manage (less consistency issues from multiple indices).
> > ```
> 
> The idea is to put all state related to an detector to the info index. As you said, additional detector info field will be added to the info index if required.

I think we should weigh the two options carefully. 1. creating a new index and 2. adding a new field to existing index. From what I have seen, option #2 is the more commonly used pattern in es and it's more efficient and easier to manage.

wnbts commented:
> > 2. After initialization is completed, the status is changed and collecting data for computing initializatio progress should stop. Collecting data and computing can be done on request instead.
> > ```
> 
> I prefer to keep collecting after the status is changed:
> First, we need extra logic to keep track of detector state. For example, a detector is stopped and its init_progress should be recalculated (we don't know its models are deleted or not, its detector definition is updated or not). Tracking the state is not trivial, and we should not put the burden on ModelManager or AnomalyResultTransportAction.
> Second, tracking can help us debugging when we suspect whether the model is not working or not.

Tracking the progress can be done on request when the get profile api is called, correct?

kaituo commented:
> > > Two general comments.
> > > ```
> > > 1. Does additional detector info require an additional data index? An additional internal field in detector is a better option for it requires less resources (no new index or redundant data fields) and easier to manage (less consistency issues from multiple indices).
> > > ```
> > 
> > 
> > The idea is to put all state related to an detector to the info index. As you said, additional detector info field will be added to the info index if required.
> 
> I think we should weigh the two options carefully. 1. creating a new index and 2. adding a new field to existing index. From what I have seen, option #2 is the more commonly used pattern in es and it's more efficient and easier to manage.

We should use an index based on its usage.  Detector index is for detector definition, not intermediate state.  That's why we have checkpoint, job, result index.  Currently, info index is created to store intermediate state, not just init_progress related.  For example, we can implement custom routing and store the routing table in the info index.  Custom routing allows more flexible routing like allocating a model on a node with the most memory available.

kaituo commented:
> > > 1. After initialization is completed, the status is changed and collecting data for computing initializatio progress should stop. Collecting data and computing can be done on request instead.
> > 
> > 
> > I prefer to keep collecting after the status is changed:
> > First, we need extra logic to keep track of detector state. For example, a detector is stopped and its init_progress should be recalculated (we don't know its models are deleted or not, its detector definition is updated or not). Tracking the state is not trivial, and we should not put the burden on ModelManager or AnomalyResultTransportAction.
> > Second, tracking can help us debugging when we suspect whether the model is not working or not.
> 
> Tracking the progress can be done on request when the get profile api is called, correct?

That requires a broadcast call to all nodes since we don't know which node has which model.  Kibana would call this api every few minutes, which is costly.

wnbts commented:
> > > > Two general comments.
> > > > ```
> > > > 1. Does additional detector info require an additional data index? An additional internal field in detector is a better option for it requires less resources (no new index or redundant data fields) and easier to manage (less consistency issues from multiple indices).
> > > > ```
> > > 
> > > 
> > > The idea is to put all state related to an detector to the info index. As you said, additional detector info field will be added to the info index if required.
> > 
> > 
> > I think we should weigh the two options carefully. 1. creating a new index and 2. adding a new field to existing index. From what I have seen, option #2 is the more commonly used pattern in es and it's more efficient and easier to manage.
> 
> We should use an index based on its usage. Detector index is for detector definition, not intermediate state. That's why we have checkpoint, job, result index. Currently, info index is created to store intermediate state, not just init_progress related. For example, we can implement custom routing and store the routing table in the info index. Custom routing allows more flexible routing like allocating a model on a node with the most memory available.

is the new index certainly needed? if detector index is to keep unchanged, what about the job index so all the execution related details/job info is managed in one index rather than scattered across multiple indices?

wnbts commented:
> > > > 1. After initialization is completed, the status is changed and collecting data for computing initializatio progress should stop. Collecting data and computing can be done on request instead.
> > > 
> > > 
> > > I prefer to keep collecting after the status is changed:
> > > First, we need extra logic to keep track of detector state. For example, a detector is stopped and its init_progress should be recalculated (we don't know its models are deleted or not, its detector definition is updated or not). Tracking the state is not trivial, and we should not put the burden on ModelManager or AnomalyResultTransportAction.
> > > Second, tracking can help us debugging when we suspect whether the model is not working or not.
> > 
> > 
> > Tracking the progress can be done on request when the get profile api is called, correct?
> 
> That requires a broadcast call to all nodes since we don't know which node has which model. Kibana would call this api every few minutes, which is costly.

it doesn't need to be broadcast. 
Most of the time, the models are initialized, the result should be stored in the index so there is no need to recompute.

If there is no stored initialization results, the model can be located using hash or the handler just directly pulls the first partition to get the status and save it for future queries.

The idea is to only do this work when it's needed and applicable.

yizheliu-amazon commented:
> > > > Two general comments.
> > > > ```
> > > > 1. Does additional detector info require an additional data index? An additional internal field in detector is a better option for it requires less resources (no new index or redundant data fields) and easier to manage (less consistency issues from multiple indices).
> > > > ```
> > > 
> > > 
> > > The idea is to put all state related to an detector to the info index. As you said, additional detector info field will be added to the info index if required.
> > 
> > 
> > I think we should weigh the two options carefully. 1. creating a new index and 2. adding a new field to existing index. From what I have seen, option #2 is the more commonly used pattern in es and it's more efficient and easier to manage.
> 
> We should use an index based on its usage. Detector index is for detector definition, not intermediate state. That's why we have checkpoint, job, result index. Currently, info index is created to store intermediate state, not just init_progress related. For example, we can implement custom routing and store the routing table in the info index. Custom routing allows more flexible routing like allocating a model on a node with the most memory available.

In this case I think we can name the index like `.opendistro-anomaly-state`, `info` sounds very general.

kaituo commented:
> > > > > Two general comments.
> > > > > ```
> > > > > 1. Does additional detector info require an additional data index? An additional internal field in detector is a better option for it requires less resources (no new index or redundant data fields) and easier to manage (less consistency issues from multiple indices).
> > > > > ```
> > > > 
> > > > 
> > > > The idea is to put all state related to an detector to the info index. As you said, additional detector info field will be added to the info index if required.
> > > 
> > > 
> > > I think we should weigh the two options carefully. 1. creating a new index and 2. adding a new field to existing index. From what I have seen, option #2 is the more commonly used pattern in es and it's more efficient and easier to manage.
> > 
> > 
> > We should use an index based on its usage. Detector index is for detector definition, not intermediate state. That's why we have checkpoint, job, result index. Currently, info index is created to store intermediate state, not just init_progress related. For example, we can implement custom routing and store the routing table in the info index. Custom routing allows more flexible routing like allocating a model on a node with the most memory available.
> 
> is the new index certainly needed? if detector index is to keep unchanged, what about the job index so all the execution related details/job info is managed in one index rather than scattered across multiple indices?

I feel so.  We cannot modify job index as job scheduler listens to any changes in the index.  They would schedule a new job if any doc changes there.

kaituo commented:
> > > > > 1. After initialization is completed, the status is changed and collecting data for computing initializatio progress should stop. Collecting data and computing can be done on request instead.
> > > > 
> > > > 
> > > > I prefer to keep collecting after the status is changed:
> > > > First, we need extra logic to keep track of detector state. For example, a detector is stopped and its init_progress should be recalculated (we don't know its models are deleted or not, its detector definition is updated or not). Tracking the state is not trivial, and we should not put the burden on ModelManager or AnomalyResultTransportAction.
> > > > Second, tracking can help us debugging when we suspect whether the model is not working or not.
> > > 
> > > 
> > > Tracking the progress can be done on request when the get profile api is called, correct?
> > 
> > 
> > That requires a broadcast call to all nodes since we don't know which node has which model. Kibana would call this api every few minutes, which is costly.
> 
> it doesn't need to be broadcast.
> Most of the time, the models are initialized, the result should be stored in the index so there is no need to recompute.
> 
> If there is no stored initialization results, the model can be located using hash or the handler just directly pulls the first partition to get the status and save it for future queries.
> 
> The idea is to only do this work when it's needed and applicable.

Good idea.  I will do this then.  I still need the info index to store error information though.

kaituo commented:
> > > > > Two general comments.
> > > > > ```
> > > > > 1. Does additional detector info require an additional data index? An additional internal field in detector is a better option for it requires less resources (no new index or redundant data fields) and easier to manage (less consistency issues from multiple indices).
> > > > > ```
> > > > 
> > > > 
> > > > The idea is to put all state related to an detector to the info index. As you said, additional detector info field will be added to the info index if required.
> > > 
> > > 
> > > I think we should weigh the two options carefully. 1. creating a new index and 2. adding a new field to existing index. From what I have seen, option #2 is the more commonly used pattern in es and it's more efficient and easier to manage.
> > 
> > 
> > We should use an index based on its usage. Detector index is for detector definition, not intermediate state. That's why we have checkpoint, job, result index. Currently, info index is created to store intermediate state, not just init_progress related. For example, we can implement custom routing and store the routing table in the info index. Custom routing allows more flexible routing like allocating a model on a node with the most memory available.
> 
> In this case I think we can name the index like `.opendistro-anomaly-state`, `info` sounds very general.

We can do that.  Yaliang may add sth in the index for his workbench.  Will collect his idea on the name as well.

wnbts commented:
> > > > > > Two general comments.
> > > > > > ```
> > > > > > 1. Does additional detector info require an additional data index? An additional internal field in detector is a better option for it requires less resources (no new index or redundant data fields) and easier to manage (less consistency issues from multiple indices).
> > > > > > ```
> > > > > 
> > > > > 
> > > > > The idea is to put all state related to an detector to the info index. As you said, additional detector info field will be added to the info index if required.
> > > > 
> > > > 
> > > > I think we should weigh the two options carefully. 1. creating a new index and 2. adding a new field to existing index. From what I have seen, option #2 is the more commonly used pattern in es and it's more efficient and easier to manage.
> > > 
> > > 
> > > We should use an index based on its usage. Detector index is for detector definition, not intermediate state. That's why we have checkpoint, job, result index. Currently, info index is created to store intermediate state, not just init_progress related. For example, we can implement custom routing and store the routing table in the info index. Custom routing allows more flexible routing like allocating a model on a node with the most memory available.
> > 
> > 
> > is the new index certainly needed? if detector index is to keep unchanged, what about the job index so all the execution related details/job info is managed in one index rather than scattered across multiple indices?
> 
> I feel so. We cannot modify job index as job scheduler listens to any changes in the index. They would schedule a new job if any doc changes there.

ok. a new index it is then.

one minor thing regarding naming, i in general avoid names like info/data for obvious reasons in software. state is good, or i wouldn't mind a name like internal state to be clear and explicit.

kaituo commented:
> > > > > > > Two general comments.
> > > > > > > ```
> > > > > > > 1. Does additional detector info require an additional data index? An additional internal field in detector is a better option for it requires less resources (no new index or redundant data fields) and easier to manage (less consistency issues from multiple indices).
> > > > > > > ```
> > > > > > 
> > > > > > 
> > > > > > The idea is to put all state related to an detector to the info index. As you said, additional detector info field will be added to the info index if required.
> > > > > 
> > > > > 
> > > > > I think we should weigh the two options carefully. 1. creating a new index and 2. adding a new field to existing index. From what I have seen, option #2 is the more commonly used pattern in es and it's more efficient and easier to manage.
> > > > 
> > > > 
> > > > We should use an index based on its usage. Detector index is for detector definition, not intermediate state. That's why we have checkpoint, job, result index. Currently, info index is created to store intermediate state, not just init_progress related. For example, we can implement custom routing and store the routing table in the info index. Custom routing allows more flexible routing like allocating a model on a node with the most memory available.
> > > 
> > > 
> > > is the new index certainly needed? if detector index is to keep unchanged, what about the job index so all the execution related details/job info is managed in one index rather than scattered across multiple indices?
> > 
> > 
> > I feel so. We cannot modify job index as job scheduler listens to any changes in the index. They would schedule a new job if any doc changes there.
> 
> ok. a new index it is then.
> 
> one minor thing regarding naming, i in general avoid names like info/data for obvious reasons in software. state is good, or i wouldn't mind a name like internal state to be clear and explicit.

What're the "obvious reasons in software"?

wnbts commented:
> > > > > > > > Two general comments.
> > > > > > > > ```
> > > > > > > > 1. Does additional detector info require an additional data index? An additional internal field in detector is a better option for it requires less resources (no new index or redundant data fields) and easier to manage (less consistency issues from multiple indices).
> > > > > > > > ```
> > > > > > > 
> > > > > > > 
> > > > > > > The idea is to put all state related to an detector to the info index. As you said, additional detector info field will be added to the info index if required.
> > > > > > 
> > > > > > 
> > > > > > I think we should weigh the two options carefully. 1. creating a new index and 2. adding a new field to existing index. From what I have seen, option #2 is the more commonly used pattern in es and it's more efficient and easier to manage.
> > > > > 
> > > > > 
> > > > > We should use an index based on its usage. Detector index is for detector definition, not intermediate state. That's why we have checkpoint, job, result index. Currently, info index is created to store intermediate state, not just init_progress related. For example, we can implement custom routing and store the routing table in the info index. Custom routing allows more flexible routing like allocating a model on a node with the most memory available.
> > > > 
> > > > 
> > > > is the new index certainly needed? if detector index is to keep unchanged, what about the job index so all the execution related details/job info is managed in one index rather than scattered across multiple indices?
> > > 
> > > 
> > > I feel so. We cannot modify job index as job scheduler listens to any changes in the index. They would schedule a new job if any doc changes there.
> > 
> > 
> > ok. a new index it is then.
> > one minor thing regarding naming, i in general avoid names like info/data for obvious reasons in software. state is good, or i wouldn't mind a name like internal state to be clear and explicit.
> 
> What're the "obvious reasons in software"?

in software, code directs data. one should be surprised if a domain term is not data/info. Consider a book object vs a book data object. there is no information gain in stating the obvious.

kaituo commented:
> > > > > > Two general comments.
> > > > > > ```
> > > > > > 1. Does additional detector info require an additional data index? An additional internal field in detector is a better option for it requires less resources (no new index or redundant data fields) and easier to manage (less consistency issues from multiple indices).
> > > > > > ```
> > > > > 
> > > > > 
> > > > > The idea is to put all state related to an detector to the info index. As you said, additional detector info field will be added to the info index if required.
> > > > 
> > > > 
> > > > I think we should weigh the two options carefully. 1. creating a new index and 2. adding a new field to existing index. From what I have seen, option #2 is the more commonly used pattern in es and it's more efficient and easier to manage.
> > > 
> > > 
> > > We should use an index based on its usage. Detector index is for detector definition, not intermediate state. That's why we have checkpoint, job, result index. Currently, info index is created to store intermediate state, not just init_progress related. For example, we can implement custom routing and store the routing table in the info index. Custom routing allows more flexible routing like allocating a model on a node with the most memory available.
> > 
> > 
> > In this case I think we can name the index like `.opendistro-anomaly-state`, `info` sounds very general.
> 
> We can do that. Yaliang may add sth in the index for his workbench. Will collect his idea on the name as well.

Changed to .opendistro-anomaly-state.

ylwu-amzn commented:
> > > > > > > Two general comments.
> > > > > > > ```
> > > > > > > 1. Does additional detector info require an additional data index? An additional internal field in detector is a better option for it requires less resources (no new index or redundant data fields) and easier to manage (less consistency issues from multiple indices).
> > > > > > > ```
> > > > > > 
> > > > > > 
> > > > > > The idea is to put all state related to an detector to the info index. As you said, additional detector info field will be added to the info index if required.
> > > > > 
> > > > > 
> > > > > I think we should weigh the two options carefully. 1. creating a new index and 2. adding a new field to existing index. From what I have seen, option #2 is the more commonly used pattern in es and it's more efficient and easier to manage.
> > > > 
> > > > 
> > > > We should use an index based on its usage. Detector index is for detector definition, not intermediate state. That's why we have checkpoint, job, result index. Currently, info index is created to store intermediate state, not just init_progress related. For example, we can implement custom routing and store the routing table in the info index. Custom routing allows more flexible routing like allocating a model on a node with the most memory available.
> > > 
> > > 
> > > In this case I think we can name the index like `.opendistro-anomaly-state`, `info` sounds very general.
> > 
> > 
> > We can do that. Yaliang may add sth in the index for his workbench. Will collect his idea on the name as well.
> 
> Changed to .opendistro-anomaly-state.

Sudipto has some comments on AD workbench, will discuss with him today. How about we split this big code change into several smaller ones? I think we need more discussion about the new index. How about we make changes to support progress bar based on current AD result index to mitigate long initialization problem first?

kaituo commented:
> > > > > > > > Two general comments.
> > > > > > > > ```
> > > > > > > > 1. Does additional detector info require an additional data index? An additional internal field in detector is a better option for it requires less resources (no new index or redundant data fields) and easier to manage (less consistency issues from multiple indices).
> > > > > > > > ```
> > > > > > > 
> > > > > > > 
> > > > > > > The idea is to put all state related to an detector to the info index. As you said, additional detector info field will be added to the info index if required.
> > > > > > 
> > > > > > 
> > > > > > I think we should weigh the two options carefully. 1. creating a new index and 2. adding a new field to existing index. From what I have seen, option #2 is the more commonly used pattern in es and it's more efficient and easier to manage.
> > > > > 
> > > > > 
> > > > > We should use an index based on its usage. Detector index is for detector definition, not intermediate state. That's why we have checkpoint, job, result index. Currently, info index is created to store intermediate state, not just init_progress related. For example, we can implement custom routing and store the routing table in the info index. Custom routing allows more flexible routing like allocating a model on a node with the most memory available.
> > > > 
> > > > 
> > > > In this case I think we can name the index like `.opendistro-anomaly-state`, `info` sounds very general.
> > > 
> > > 
> > > We can do that. Yaliang may add sth in the index for his workbench. Will collect his idea on the name as well.
> > 
> > 
> > Changed to .opendistro-anomaly-state.
> 
> Sudipto has some comments on AD workbench, will discuss with him today. How about we split this big code change into several smaller ones? I think we need more discussion about the new index. How about we make changes to support progress bar based on current AD result index to mitigate long initialization problem first?

We cannot support progress bar based on current AD result index.  What's the undecided besides index name?  We can easily change name for that.

What's an easy way to split this?  Half of the PR are tests.  I have mainly three parts: one is the change to store to new index , one is new transport action to support fetching rcf updates, another part is to changes in the consumer (profile api).  If I split in this way, one PR has roughly 1000 lines, another PR has 600 lines, another PR has 1400 lines.  Are you good with that?

wnbts commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/indices/AnomalyDetectionIndices.java:
```diff
@@ -229,6 +251,18 @@ public void initAnomalyDetectorJobIndex(ActionListener<CreateIndexResponse> acti
+     * Create an index.
+     *
+     * @param actionListener action called after create index
+     * @throws IOException IOException from {@link AnomalyDetectionIndices#getAnomalyDetectorJobMappings}
```
remark. the issue of checked exception is clear here. they bubble up from low levels and raise unhelpful concerns/threats to higher-level in the chain, which, in this case, choose to expose unnecessary implementation.

when there is an unlikely checked exception due to lower-level implementation and the caller is unlikely to act or recover, it's likely a good place to wrap the checked exceptions early on.

kaituo replied:
I did it. See ThrowingConsumerWrapper.throwingConsumerWrapper.

wnbts commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/ml/ModelManager.java:
```diff
@@ -1039,4 +1039,18 @@ private double computeRcfConfidence(RandomCutForest forest) {
+      * @param modelId model id
+      */
+    public long getTotalUpdates(String modelId) {
+        ModelState<RandomCutForest> model = forests.get(modelId);
```
issue. what happens to a completed but not hosted model (maybe after detector is stopped)?

kaituo replied:
We return 0 in this case.  Any other better way?

wnbts replied:
the code should load the checkpoint and return its count.

kaituo replied:
make sense

ylwu-amzn replied:
I think we only show progress bar when detector started. Once detector start, will restore model from checkpoint. Then we can read updates from hosted model. Should we pull the checkpoint again?

kaituo replied:
we pull checkpoint when we cannot get updates from hosted models.  We may not do this if use the state index.

ylwu-amzn replied:
Should we show progress before restoring model? For example, one detector has 10 minutes interval, when restart detector, we show progress 80% directly as we pull checkpoint (or state index)? Is it possible we fail to restore model and the progress still show as 80%?

kaituo replied:
We need to pull checkpoint when shingle is not ready and model is in memory yet.  Sometimes this process is long, sometimes is short.  This is a tradeoff between code simplicity (maintenance/testability) vs accuracy.  What you suggested is what I did one week ago.  Now I changed my code to more complex and more accurate side.  There is no absolute right or wrong.  Let's stick to current complex solution.

kaituo replied:
I thought about it during the vacation.  I plan to use messaging + checkpoint polling, and remove rcf updates on index.  Lai's suggested "so when the count is above 128, the handler just returns it without doing anything extra".  Messaging + checkpoint polling is equally efficient with getting document from an index, if not faster.  And we save the trouble of maintaining another field on the state index.  Let me know if you guys think differently.

wnbts commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/model/InitProgressProfile.java:
```diff
@@ -0,0 +1,147 @@
+ *   During initialization, we need to tell users progress (using a percentage), how many more
+ *    shingles to go, and approximately how many minutes before the detector becomes operational
+ *     if they keep their data stream continuous.
+ * @author kaituo
```
minor. unneeded annotation.

kaituo replied:
removed

wnbts commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/model/InitProgressProfile.java:
```diff
@@ -0,0 +1,147 @@
+/**
+ * Profile output for detector initialization progress.  When the new detector is created, it is possible that
+ * there hasn’t been enough continuous data in the index.  We need to use live data to initialize.
+ *   During initialization, we need to tell users progress (using a percentage), how many more
+ *    shingles to go, and approximately how many minutes before the detector becomes operational
+ *     if they keep their data stream continuous.
```
minor. indentation is off.

kaituo replied:
fixed.

kaituo commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/ml/ModelManager.java:
```diff
@@ -1039,4 +1039,18 @@ private double computeRcfConfidence(RandomCutForest forest) {
     }
+
+    /**
+      * Get all RCF partition's size corresponding to a detector.  Thresholding
+      *  models' size is a constant since they are small in size (KB).
```
good catch. Fixed.

ylwu-amzn replied:
> good catch. Fixed.

Seems new revision not published yet?

kaituo replied:
yeah, I am waiting to reach an agreement with Lai: whether to use index state or transport + checkpoint pull.  I am ok with both.

kaituo commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/model/DetectorInternalState.java:
```diff
@@ -0,0 +1,186 @@
+        try {
+            state = (DetectorInternalState) super.clone();
+        } catch (CloneNotSupportedException e) {
+            state = new DetectorInternalState.Builder().rcfUpdates(rcfUpdates).lastUpdateTime(lastUpdateTime).error(error).build();
```
fixed

kaituo commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/ADStateManager.java:
```diff
@@ -52,6 +53,7 @@
     private static final Logger LOG = LogManager.getLogger(ADStateManager.class);
     private ConcurrentHashMap<String, Entry<AnomalyDetector, Instant>> currentDetectors;
     private ConcurrentHashMap<String, Entry<Integer, Instant>> partitionNumber;
+    private ConcurrentHashMap<String, Instant> currentCheckpoints;
```
Could you explain your question?

This is for starting cold start whenever there is missing checkpoint.  The key is detector id, while the value is the timestamp when we find a checkpoint.

wnbts replied:
there are three maps, if their keys are detector ids, they can be merged into one map sharing the same keys.

kaituo replied:
How do I merge it?  Their values are different.

wnbts replied:
a manager should manage a set of assigned detectors so for an assigned detector id, there should be one key-vale pair of detector id, run state. the various parts can be put together into one structure rather than requiring separate key-value pairs for each field.

kaituo replied:
will create a class ADState then.

kaituo commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/ADStateManager.java:
```diff
@@ -140,6 +143,30 @@ public void getAnomalyDetector(String adID, ActionListener<Optional<AnomalyDetec
         }, listener::onFailure);
     }
 
+    public void getDetectorCheckpoint(String adID, ActionListener<Boolean> listener) {
```
doc is deprecated.  We should remove doc in checkpointdao.  Do you want me to do that?

wnbts replied:
i mean the doc for this getDetectorCheckpoint method is missing. a client/reader cannot know what this method does without reading through code.

kaituo replied:
added doc

kaituo commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/ADStateManager.java:
```diff
@@ -176,6 +205,27 @@ public void maintenance() {
+     *
+     * @param flags flags to be maintained
+     */
+    void maintenanceFlag(ConcurrentHashMap<String, Instant> flags) {
```
I wrote unit tests for it.  That's why I am not using private.

wnbts replied:
try to avoid a. leaking implementation in class interface. and b. build brittle implementation-based tests.

kaituo replied:
duly noted.  Having package private for tests is not uncommon.  People invented @VisibleForTesting annotation for that.

wnbts replied:
that's better left for legacy code. for new code, it's better not to start with it.

kaituo commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/AnomalyResultTransportAction.java:
```diff
@@ -287,6 +291,20 @@ protected void doExecute(Task task, ActionRequest actionRequest, ActionListener<
             }
 
             if (!featureOptional.getProcessedFeatures().isPresent()) {
```
other reasons could be: circuit breaker is open, cluster is under read/write block, or AD is disabled.  We should not trigger cold start in those cases.

wnbts replied:
ok. it might be good to explicitly put in code the conditions where cold start is suppressed so that no one needs to guess why initialization is not done

kaituo replied:
added doc

kaituo commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/handler/DetectorStateHandler.java:
```diff
@@ -0,0 +1,175 @@
+        );
+    }
+
+    public void saveRcfUpdates(long totalRcfUpdates, String detectorId) {
```
make sense.  I'll modify when you answered my previous question on whether to continue to use on-demand messaging.  Also, we should use 128, right?

wnbts replied:
128 is the current number

kaituo commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/util/ThrowingConsumerWrapper.java:
```diff
@@ -0,0 +1,41 @@
+
+public class ThrowingConsumerWrapper {
+    /**
+     * Utility method to use a method throwing checked exception inside a function
```
Could you explain your remark?

wnbts replied:
if underlying low-level code doesn't threaten with checked exceptions, this wouldn't be necessary, right?

kaituo replied:
yes

wnbts commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/AnomalyDetectorJobRunner.java:
```diff
@@ -436,7 +442,8 @@ private void indexAnomalyResult(
             );
-            anomalyResultHandler.indexAnomalyResult(anomalyResult);
+            anomalyResultHandler.index(anomalyResult, detectorId);
+            detectorStateHandler.saveError(response.getError(), detectorId);
```
same idea applies to error and count, a write only happens if there is an update to the read. if the error is null/the same, it is not needed to update.

regarding the last question, i suggest writes (due to the count field) only happen when the count is below 128 and new count is different (larger) than the old count. Messaging is one way to obtain the current count but i think loading the checkpoint locally to get the persisted count also suffices.

kaituo replied:
Error needs to be up-to-date.  When a profile request comes in, it won't know this error is up-to-date or not unless the writer side guarantees it.  For example, if there is no error, a profile API's request comes in to ask for error, we would return an error instead of null.

About the in-progress number, how about the we always depending on on-demand messaging.  If there is rcf model in memory, we ask for it.  Otherwise, we ask for checkpoint.  We can remove index write for rcf updates.

wnbts replied:
regarding error, the main idea is to update only when it's changed. for example, error to null/non-error, null to error, or error 1 to error 2. if it's null to null, error 1 to error 1, no update is needed.

for in-progress number, i would like to see the info is obtained on demand to make sure it's current. two ramifications. first, if the state (count >= 128) is final, a query to index should be better than a transport + a checkpoint pull from index. second, if progress doesn't depend on index, does error alone justify index?

kaituo replied:
Updating error when it is changed works.  Will change.

about your ramifications: 
1) if we update when it is less than 128, why do we still need a transport + a checkpoint pull?
2) yes, Yaliang and folks will add things there as needed.

wnbts replied:
1. when the count in the previously saved state is in [0, 128), in order to get the current count to a. return in response and b. update state with, there is two ways: 1. transport + checkpoint or 2. just checkpoint. (either is ok to me, latter is simpler, more reliable but slower to change, hourly or so).

when the count in the saved state is no less than 128, it is returned in response and done. transport or checkpoint query is no longer needed.

ylwu-amzn commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/AnomalyDetectorPlugin.java:
```diff
@@ -170,28 +175,34 @@ public AnomalyDetectorPlugin() {}
     ) {
-        AnomalyResultHandler anomalyResultHandler = new AnomalyResultHandler(
+
+        AnomalyIndexHandler<AnomalyResult> anomalyResultHandler;
```
minor, can we rename `anomalyResultHandler ` as `anomalyIndexHandler`?

kaituo replied:
anomalyResultHandler seems more appropriate than anomalyIndexHandler since this handler is for anomaly result.

ylwu-amzn replied:
How about we name the class as `AnomalyResultHandler` if it's for anomaly result only?

kaituo replied:
The class can be inherited by other classes writing to a different index.  We shouldn't do that.

ylwu-amzn commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/model/DetectorInternalState.java:
```diff
@@ -0,0 +1,186 @@
+        it -> parse(it)
+    );
+
+    public static final String DETECTOR_STATE_INDEX = ".opendistro-anomaly-state";
```
How about change to `.opendistro-anomaly-detection-state`?

kaituo replied:
changed

ylwu-amzn commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/model/DetectorInternalState.java:
```diff
@@ -0,0 +1,186 @@
+    public static final String LAST_UPDATE_TIME_FIELD = "last_update_time";
+    public static final String ERROR_FIELD = "error";
+
+    private long rcfUpdates = -1L;
```
Any special consideration for using `-1` as default value?

kaituo replied:
want to make it a illegal value. 0 is legal value.

ylwu-amzn commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/ADStateManager.java:
```diff
@@ -148,11 +175,13 @@ public void getAnomalyDetector(String adID, ActionListener<Optional<AnomalyDetec
     public void maintenance() {
         maintenance(currentDetectors);
         maintenance(partitionNumber);
+        maintenanceFlag(currentCheckpoints);
```
Why need to periodically maintain the checkpoint flag?

kaituo replied:
otherwise, who would maintain the flag?

ylwu-amzn replied:
Sorry, not explain clearly. My question is why need "periodically" maintain. Can we just maintain it when update/delete happen?

kaituo replied:
make sense.  We both need to remember to do this when we fix checkpoint deletion.

kaituo replied:
I remember why I did it last year: this adds fault tolerance for state cleanup.  For example, what if customers never delete checkpoint and keep the detector stopped.  Then the checkpoint state is kept there in limbo in memory forever.

ylwu-amzn commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/ADStateManager.java:
```diff
@@ -140,6 +143,30 @@ public void getAnomalyDetector(String adID, ActionListener<Optional<AnomalyDetec
+    private ActionListener<GetResponse> onGetCheckpointResponse(String adID, ActionListener<Boolean> listener) {
+        return ActionListener.wrap(response -> {
+            if (response == null || !response.isExists()) {
+                listener.onResponse(Boolean.FALSE);
```
If not find checkpoint, we should remove it from `currentCheckpoints `.

Another case is if user delete checkpoint manually, we should remove it from `currentCheckpoints`.

kaituo replied:
We would come to this code path if currentCheckpoints has the flag set in the first place.

Currently, we don't delete checkpoint.  I am waiting for you to fix it :)

ylwu-amzn replied:
> We wouldn't come to this code path if currentCheckpoints has the flag set in the first place.
> 
> Currently, we don't delete checkpoint. I am waiting for you to fix it :)

Sure, can you add some todo here?

kaituo replied:
Created https://github.com/opendistro-for-elasticsearch/anomaly-detection/issues/183

ylwu-amzn commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/AnomalyResultTransportAction.java:
```diff
@@ -287,6 +291,20 @@ protected void doExecute(Task task, ActionRequest actionRequest, ActionListener<
             }
 
             if (!featureOptional.getProcessedFeatures().isPresent()) {
+                stateManager.getDetectorCheckpoint(adID, ActionListener.wrap(checkpointExists -> {
```
For happy case(detector and feature present and circuit breaker not open, no read/write block, AD enabled), will every job run call `getDetectorCheckpoint ` method?

kaituo replied:
yes

ylwu-amzn replied:
AD will take snapshot of model and save as checkpoint every 1 hour? If one detector has 1 minute interval and no checkpoint when start, is it possible that first 59 detector runs can't get checkpoint and do cold start?

kaituo replied:
we also take checkpoint during cold start.  So in your example, we have one checkpoint after the 1st run and the next one is taken on the 61st run.

ylwu-amzn replied:
Cool, that makes sense.

ylwu-amzn commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/RCFPollingTransportAction.java:
```diff
@@ -0,0 +1,130 @@
+import com.amazon.opendistroforelasticsearch.ad.ml.ModelManager;
+import com.amazon.opendistroforelasticsearch.ad.settings.AnomalyDetectorSettings;
+
+public class RCFPollingTransportAction extends HandledTransportAction<RCFPollingRequest, RCFPollingResponse> {
```
Add some comment doc for this class?

kaituo replied:
added

ylwu-amzn commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/ADStateManager.java:
```diff
@@ -176,6 +205,27 @@ public void maintenance() {
+            try {
+                Instant time = entry.getValue();
+                if (time.plus(stateTtl).isBefore(clock.instant())) {
+                    flags.remove(detectorId);
```
We can save some memory if remove detector from `currentCheckpoints`, but needs another index query to get checkpoint as almost every detector run will try to get checkpoint. Can we just remove it when detector stopped/deleted?

kaituo replied:
state manitenance would remove it from memory.  We don't have to do it manually.

ylwu-amzn replied:
> state manitenance would remove it from memory. We don't have to do it manually.

Yeah, the question is should we add some TTL to checkpoint cache and remove it periodically? Can we simplify the logic to just remove one detector's checkpoint cache when detector stopped/deleted?

kaituo replied:
One case this periodic cleanup is useful: job schedule's shard relocate and we don't use the coordinating node anymore.  Before the detector gets deleted/stopped, these memory would be there.

wnbts commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/handler/DetectorStateHandler.java:
```diff
@@ -0,0 +1,163 @@
+        try {
+            GetRequest getRequest = new GetRequest(this.indexName).id(detectorId);
+
+            clientUtil.<GetRequest, GetResponse>asyncRequest(getRequest, client::get, ActionListener.wrap(response -> {
+                DetectorInternalState newState = null;
```
can error be part of transport state and trigger indexing when updated? so a get can be saved for each run.

kaituo replied:
On second thought, what you said make sense.  I'll keep the last error in transport state and access it in DetectorStateHandler.

wnbts replied:
if after each run, the detector coordinator itself, holding the anomaly result and error, manages detector error state rather than letting the job scheduler do it, does it still require messaging?

kaituo replied:
You meant does it still require indexing into state index?

Yes, the state can be cleared due to maintenance, or coordinating node can change.  We need a permanent place to host the data.

wnbts commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/TransportStateManager.java:
```diff
@@ -140,35 +149,61 @@ public void getAnomalyDetector(String adID, ActionListener<Optional<AnomalyDetec
+     * @param adID  the detector's ID
+     * @param listener listener to handle get request
+     */
+    public void getDetectorCheckpoint(String adID, ActionListener<Boolean> listener) {
```
found it

kaituo commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/AnomalyDetectorProfileRunner.java:
```diff
@@ -203,242 +190,120 @@ private void prepareProfile(
-                        detectorId
-                    );
-                listener.onFailure(new RuntimeException("Fail to find detector state: " + detectorId, exception));
+            if (profilesToCollect.contains(ProfileName.INIT_PROGRESS)) {
+                listener.onResponse(new DetectorProfile.Builder().build());
```
This means init progress is empty for a stopped detector.

wnbts replied:
if the status of the model is completed, i would think the profile init status should also be completed to reflect the truth, whether the detector is running or stopped. otherwise, it might cause confusion that stopping a detector deletes the model or uncertainty about the model of a stopped detector.

kaituo commented on src/main/resources/mappings/anomaly-state.json:
```diff
@@ -0,0 +1,21 @@
+    "schema_version": {
+      "type": "integer"
+    },
+    "rcf_updates": {
+      "type": "integer"
+    },
```
good catch.  Removed

kaituo commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/ml/ModelManager.java:
```diff
@@ -1045,4 +1067,24 @@ private double computeRcfConfidence(RandomCutForest forest) {
+        ModelState<RandomCutForest> model = forests.get(modelId);
+        if (model != null) {
+            listener.onResponse(model.getModel().getTotalUpdates());
+        } else {
+            checkpointDao
+                .getModelCheckpoint(
```
yeah, fortunately this is one-time thing and we don't have to do it repeatedly.  Let me know if you any way to improve this.

wnbts replied:
i am worried about for every get profile call, the model gets pulled, loaded, and hosted.

some optimizations
1. there is not much to save from getting the checkpoint. but the checkpoint doesn't need to be fully deserialized, i.e., the bulk of the model, the trees are not needed. a lightweight deserialization only involving the forest containing the update count can be added to avoid recomputing trees and discarding them.
2. if the model is not hosted before deserialization, it doesn't need to be hosted since it's not being used by active jobs, saves memory space and write back to index again later during maintenance.

kaituo replied:
The model does not need to be pulled, loaded, and hosted for every get profile call.  Only the first profile call if the model hasn't been hosted.

kaituo commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/AnomalyDetectorProfileRunner.java:
```diff
@@ -203,242 +190,120 @@ private void prepareProfile(
+                    long lastUpdateTimeMs = detectorState.getLastUpdateTime().toEpochMilli();
+
+                    // if state index hasn't been updated, we should not use the error field
+                    // For example, before enabled, error is stopped due to, we should not show
```
Is the following easier to read?

                    // if state index hasn't been updated, we should not use the error field
                    // For example, before enabled, error is "stopped due to blah", we should not show
                    // this when the detector is enabled.

wnbts replied:
ok. now i understand it. a more complete sentence will be clearer.

For example, before (the detector is) enabled, if the error message contains the phrase "stopped due to blah", ....

kaituo replied:
will change.

wnbts commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/TransportStateManager.java:
```diff
@@ -91,31 +92,31 @@ public ADStateManager(
     public int getPartitionNumber(String adID, AnomalyDetector detector) {
-        Entry<Integer, Instant> partitonAndTime = partitionNumber.get(adID);
-        if (partitonAndTime != null) {
+        if (transportStates.containsKey(adID) && transportStates.get(adID).getPartitonNumber() != null) {
```
minor. computeIfPresent is more concise

wnbts commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/TransportStateManager.java:
```diff
@@ -130,7 +131,13 @@ public void getAnomalyDetector(String adID, ActionListener<Optional<AnomalyDetec
                 ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.nextToken(), parser::getTokenLocation);
                 AnomalyDetector detector = AnomalyDetector.parse(parser, response.getId());
-                currentDetectors.put(adID, new SimpleEntry<>(detector, clock.instant()));
+                TransportState state = transportStates.get(adID);
```
minor. computeIfAbsent is more concise.

wnbts commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/TransportStateManager.java:
```diff
@@ -204,4 +237,29 @@ public void resetBackpressureCounter(String nodeId) {
+            return errorAndTime.getKey();
+        }
+
+        return IMPOSSIBLE_ERROR;
```
minor. this seems more no_error.

ylwu-amzn commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/transport/handler/DetectorStateHandler.java:
```diff
@@ -0,0 +1,165 @@
+
+    public void saveError(String error, String detectorId) {
+        // trigger indexing if no error recorded (e.g., this detector got enabled just now)
+        // or the recorded error is different than this one.
```
Should we update error to record the latest error time if the error message is the same?

kaituo replied:
We need last update time to know if the error is recorded before or after enabled time.  Current code path can handle that.  So we don't need to record the latest error time if the error message is the same.

ylwu-amzn commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/settings/AnomalyDetectorSettings.java:
```diff
@@ -141,6 +141,7 @@ private AnomalyDetectorSettings() {}
     public static final String ANOMALY_DETECTORS_INDEX_MAPPING_FILE = "mappings/anomaly-detectors.json";
     public static final String ANOMALY_DETECTOR_JOBS_INDEX_MAPPING_FILE = "mappings/anomaly-detector-jobs.json";
     public static final String ANOMALY_RESULTS_INDEX_MAPPING_FILE = "mappings/anomaly-results.json";
+    public static final String ANOMALY_DETECTOR_STATE_INDEX_MAPPING_FILE = "mappings/anomaly-state.json";
```
Minor, how about rename  the json file as "anomaly-detection-state.json"? In AD workbench, we will introduce detection task, plan to put task execution state in this index ".opendistro-anomaly-detection-state". 
BTW, how about we rename ANOMALY_DETECTOR_STATE_INDEX_MAPPING_FILE as ANOMALY_DETECTION_STATE_INDEX_MAPPING_FILE, similar to ANOMALY_DETECTOR_STATE_INDEX.

kaituo replied:
changed both

wnbts commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/AnomalyDetectorJobRunner.java:
```diff
@@ -436,7 +442,8 @@ private void indexAnomalyResult(
             );
-            anomalyResultHandler.indexAnomalyResult(anomalyResult);
+            anomalyResultHandler.index(anomalyResult, detectorId);
+            detectorStateHandler.saveError(response.getError(), detectorId);
```
1. if job runner (sender) and state manager (responder) are on separate nodes for better scalability in the future, why let this change add an unnecessary coupling to prevent that?

2. the state manager is already a dependency of transport action, which generates errors first hand, why not letting the transport action directly update state manager, which updates the handler, (the chain of command is job runner > transport action > state manager > state handler), rather than taking a longer route involving the high-level job runner for more lower level details, which should only be responsible for sending job requests by schedule and have as little business logic as possible?

kaituo replied:
1. I don't see it is coming in the near feature.  State manager is saving the state of job runner.  It makes no sense to save it in different node.
2. detectorStateHandler and anomalyResultHandler are both dependency of job runner.  Previously, the logic of anomalyResultHandler is in the transport action, as you suggested.  Yaliang separated it out and moved it to job runner.  It is easier to read and being consistent by keeping the way as it is now.

ylwu-amzn replied:
> 1. if job runner (sender) and state manager (responder) are on separate nodes for better scalability in the future, why let this change add an unnecessary coupling to prevent that?
> 2. the state manager is already a dependency of transport action, which generates errors first hand, why not letting the transport action directly update state manager, which updates the handler, (the chain of command is job runner > transport action > state manager > state handler), rather than taking a longer route involving the high-level job runner for more lower level details, which should only be responsible for sending job requests by schedule and have as little business logic as possible?

hi, Lai, good questions, before decoupling, we have no job runner and we handle all business logic and exceptions in transport action as it stands on top of other components. After decoupling, job runner is the top component, split the exception handling to job runner to avoid the exception handling logic scattered into multiple components(transport actions, job runners, etc).  So job runner will catch all exceptions including its own exceptions, then decide the following error handling such as stop detector if no enough resource, and write error in AD result index. Now Kaituo add `detectorStateHandler`, which will track the latest error. So it's same design logic to do error handling in one place.

Of course, we may have other options to do this, like create a new error handler, pass it to every component to handle exception and persist in AD result individually. Both options have pros and cons, need to do some tradeoff. How about we keep current design for this PR? We can discuss different options and solve the error handling in another PR if we have better option.

wnbts replied:
this is not a blocker, as i have approved the pr.

it's good to use interface to decouple business logic from work execution dependencies. in another word, consider if the job scheduler code (a work execution dependency) change/migrate to some other dependency for execution, how much needs to re-developed, copy and pasted, re-tested? can/should the work be further reduced? in the ideal case, the interface should serve to protect the business logic.

ylwu-amzn replied:
Thanks for the suggestion. Agree to "use interface to decouple business logic from work execution dependencies".  I think we can restructure when do AD workbench as it will change a lot of how to schedule detection task/job but not overdesign. Let's discuss more when do AD workbench.

ylwu-amzn commented on src/main/java/com/amazon/opendistroforelasticsearch/ad/AnomalyDetectorJobRunner.java:
```diff
@@ -119,6 +121,10 @@ public void setSettings(Settings settings) {
         this.maxRetryForEndRunException = AnomalyDetectorSettings.MAX_RETRY_FOR_END_RUN_EXCEPTION.get(settings);
     }
 
+    public void setDetectorStateHandler(DetectorStateHandler detectorStateHandler) {
```
minor, should we rename `setDetectorStateHandler ` as `setDetectionStateHandler ` ? Same question for other `...detectorState...`

kaituo replied:
renamed
